src/transformers/models/realm/tokenization_realm.py

```diff
@@ -0,0 +1,149 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tokenization classes for REALM."""
+
+from ...file_utils import PaddingStrategy
+from ...tokenization_utils_base import BatchEncoding
+from ...utils import logging
+from ..bert.tokenization_bert import BertTokenizer
+
+
+logger = logging.get_logger(__name__)
+
+VOCAB_FILES_NAMES = {"vocab_file": "vocab.txt"}
+
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "realm-cc-news-pretrained-embedder": "https://huggingface.co/qqaatw/realm-cc-news-pretrained-embedder/resolve/main/vocab.txt",
+        "realm-cc-news-pretrained-encoder": "https://huggingface.co/qqaatw/realm-cc-news-pretrained-encoder/resolve/main/vocab.txt",
+        "realm-cc-news-pretrained-scorer": "https://huggingface.co/qqaatw/realm-cc-news-pretrained-scorer/resolve/main/vocab.txt",
+        "realm-cc-news-pretrained-openqa": "https://huggingface.co/qqaatw/realm-cc-news-pretrained-openqa/aresolve/main/vocab.txt",
+        "realm-orqa-nq-openqa": "https://huggingface.co/qqaatw/realm-orqa-nq-openqa/resolve/main/vocab.txt",
+        "realm-orqa-nq-reader": "https://huggingface.co/qqaatw/realm-orqa-nq-reader/resolve/main/vocab.txt",
+        "realm-orqa-wq-openqa": "https://huggingface.co/qqaatw/realm-orqa-wq-openqa/resolve/main/vocab.txt",
+        "realm-orqa-wq-reader": "https://huggingface.co/qqaatw/realm-orqa-wq-reader/resolve/main/vocab.txt",
+    }
+}
+
+PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES = {
+    "realm-cc-news-pretrained-embedder": 512,
+    "realm-cc-news-pretrained-encoder": 512,
+    "realm-cc-news-pretrained-scorer": 512,
+    "realm-cc-news-pretrained-openqa": 512,
+    "realm-orqa-nq-openqa": 512,
+    "realm-orqa-nq-reader": 512,
+    "realm-orqa-wq-openqa": 512,
+    "realm-orqa-wq-reader": 512,
+}
+
+PRETRAINED_INIT_CONFIGURATION = {
+    "realm-cc-news-pretrained-embedder": {"do_lower_case": True},
+    "realm-cc-news-pretrained-encoder": {"do_lower_case": True},
+    "realm-cc-news-pretrained-scorer": {"do_lower_case": True},
+    "realm-cc-news-pretrained-openqa": {"do_lower_case": True},
+    "realm-orqa-nq-openqa": {"do_lower_case": True},
+    "realm-orqa-nq-reader": {"do_lower_case": True},
+    "realm-orqa-wq-openqa": {"do_lower_case": True},
+    "realm-orqa-wq-reader": {"do_lower_case": True},
+}
+
+
+class RealmTokenizer(BertTokenizer):
+    r"""
+    Construct a REALM tokenizer.
+
+    [`RealmTokenizer`] is identical to [`BertTokenizer`] and runs end-to-end tokenization: punctuation splitting and
+    wordpiece.
+
+    Refer to superclass [`BertTokenizer`] for usage examples and documentation concerning parameters.
+    """
+
+    vocab_files_names = VOCAB_FILES_NAMES
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+    max_model_input_sizes = PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
+    pretrained_init_configuration = PRETRAINED_INIT_CONFIGURATION
+
+    def batch_encode_candidates(self, text, **kwargs):
```

Comment:
should we maybe instead overwrite the `__call__` method here? Why do we have to enforce the max-length padding strategy here?

Reply:
@LysandreJik @sgugger

Reply:
IMO this should be handled in `__call__` method.  Because this is not aligned with the tokenizer API.
For example, see the `LUKETokenizer` which overrides `__call__` to encode multiple texts 
https://github.com/huggingface/transformers/blob/master/src/transformers/models/luke/tokenization_luke.py#L262

Reply:
@patrickvonplaten The reason of requiring max-length padding strategy is that a single input would include many candidates. In the following example, a sample has two candidates, and there are two samples that subsume into a batch. The goal of this function is to produce an encoded output with `(batch_size, num_candidates, max_length)` shape for `RealmKnowledgeAugEncoder`.

```
# batch_size = 2, num_candidates = 2
batch_texts = [
    ["Hello world!", "Nice to meet you!"], # First sample
    ["The cute cat.", "The adorable dog."]  # Second sample
]

tokenized_texts = tokenizer.batch_encode_candidates(text, max_length=10, return_tensors="pt")

print(tokenized_texts.input_ids.shape) # (2, 2, 10)
```

Therefore, if we don't adopt the max-length strategy and use the longest strategy instead, based on the current logic of `batch_encode_candidates`, the first sample will be firstly inputted into `__call__` and result in an output with shape (num_candidates, longest_length_of_first_sample). Then, we do the same step for the second sample and get an output with shape (num_candidates, longest_length_of_second_sample). These two outputs cannot be stacked into a batch tensor because their dimension on the last axis is different. We could however design a more sophisticated logic to adapt the longest strategy (I can do it in this PR or in a follow-up PR).

@patrickvonplaten, @patil-suraj - This function could be integrated into `__call__`, but IMO I think keeping a separated function would be more clear and unequivocal to users i.e. using `batch_encode_candidates` when they want to encode samples contain candidates for `RealmKnowledgeAugEncoder`; otherwise, using the regular `__call__` to encode.

Reply:
I'm ok with overriding the `__call__` if you think this way is more aligned and intuitive though.